### ISSUE TITLE
Question: register two interfaces for same concrete class

### ISSUE THREAD

**Issue by dgehri** [ybainier/Hypodermic#44]
Suppose `Foo` implements `ISome` and `IThing`. How does one register both interfaces, while ensuring that there is only a single `Foo` instance? 

I can achieve it by using a factory method for one of the interfaces, and some dynamic_cast wizardry. Would be nice if Hypodermic allowed chaining `as<...>()` to register multiple interfaces.

**Comment by ybainier** [comment]
You can write
```cpp
builder.registerType< Foo >()
       .as< ISome >()
       .as< IThing >()
       .singleInstance();
```
then resolve `Foo` like so:
```cpp
auto fooAsSome = container->resolve< ISome >();
auto fooAsThing = container->resolve< IThing >();
```
both will be the same instance of `Foo`.

**Comment by dgehri** [comment]
Thanks @ybainier ! This solves the problem, but I would like to add the following, which took me some time to understand: in my case, I was registering using `builder.registerInstanceFactory(...)` where the instance factory returns a `shared_ptr<ISome>`, not a `shared_ptr<SomeObject>`, where `struct SomeObject : ISome {};` and `struct ISome : IThing {};`.

For this to work, I needed to do as follows:
```
builder.registerInstanceFactory(...)
        .as<IThing>()
        .asSelf()
        .singleInstance();
```

The order was important, and I had to use `asSelf()` instead of `as<ISome>()`.

While I understand the reasons for this, would it be possible to permit `as<ISome>()`? That is, relaxing the requirement that the template parameter to `as` be a super-class, allowing it to be the registered class itself? That would avoid a few frustrations...
